Mathlib/Tactic/TacticAnalysis/Declarations.lean

```diff
@@ -0,0 +1,135 @@
+      let name ← mkAuxDeclName `extracted
+      let ((sig, _, modules), _) ← (Mathlib.Tactic.ExtractGoal.goalSignature name goal).run
+      let imports := modules.toList.map (s!"import {·}")
+      return ([goal], m!"{"\n".intercalate imports}\n\ntheorem {sig} := by\n  fail_if_success grind\n  {stx}")
+  tell stx _old new :=
+    if new.1.1 != [] then
+      m!"'grind' failed where '{stx}' succeeded. Counterexample:\n{new.1.2}"
+    /-
+    else
+      if old.2 * 2 < new.2 then
+        return m!"'grind' is slower than '{stx}': {new.2 / 1000} versus {old.2 / 1000} heartbeats"
+    -/
+    else none }
+
+/-- Suggest merging two adjacent `rw` tactics if that also solves the goal. -/
+register_option linter.tacticAnalysis.rwMerge : Bool := {
+  defValue := false
+}
+
+@[tacticAnalysis linter.tacticAnalysis.rwMerge, inherit_doc linter.tacticAnalysis.rwMerge]
+def rwMerge : TacticAnalysis.Config := .ofComplex {
+  out := (List MVarId × Array Syntax)
+  ctx := (Array (Array Syntax))
+  trigger ctx stx :=
+    match stx with
+    | `(tactic| rw [$args,*]) => .continue ((ctx.getD #[]).push args)
+    | _ => if let some args := ctx then if args.size > 1 then .accept args else .skip else .skip
+  test ctx goal := withOptions (fun opts => opts.set `grind.warning false) do
+    let ctxT : Array (TSyntax `Lean.Parser.Tactic.rwRule) := ctx.flatten.map (⟨·⟩)
+    let tac ← `(tactic| rw [$ctxT,*])
+    try
+      let (goals, _) ← Lean.Elab.runTactic goal tac
+      return (goals, ctxT.map (↑·))
+    catch _e => -- rw throws an error if it fails to pattern-match.
+      return ([goal], ctxT.map (↑·))
+  tell _stx _old new :=
+    if new.1.1.isEmpty then
+      m!"Try this: rw {new.1.2}"
+    else none }
+
+/-- Suggest merging `tac; grind` into just `grind` if that also solves the goal. -/
+register_option linter.tacticAnalysis.mergeWithGrind : Bool := {
+  defValue := false
+}
+
+@[tacticAnalysis linter.tacticAnalysis.mergeWithGrind,
+  inherit_doc linter.tacticAnalysis.mergeWithGrind]
+def mergeWithGrind : TacticAnalysis.Config where
+  run seq := do
+    if let #[(preCtx, preI), (_postCtx, postI)] := seq[0:2].array then
+      if postI.stx.getKind == ``Lean.Parser.Tactic.grind then
+        if let [goal] := preI.goalsBefore then
+          preCtx.runTactic preI goal <| fun goal => do
+            let tac := postI.stx
+            let (goals, _) ← try
+                Lean.Elab.runTactic goal tac
+              catch _e =>
+                pure ([goal], {})
+            if goals.isEmpty then
+              logWarningAt preI.stx m!"'{preI.stx}; grind' can be replaced with 'grind'"
+
+/-- Suggest replacing a sequence of tactics with `grind` if that also solves the goal. -/
+register_option linter.tacticAnalysis.terminalToGrind : Bool := {
+  defValue := false
+}
+
+@[tacticAnalysis linter.tacticAnalysis.terminalToGrind,
+  inherit_doc linter.tacticAnalysis.terminalToGrind]
+def terminalToGrind : TacticAnalysis.Config where
```

Comment:
I don't understand how this is only testing terminal sequences of tactics. (Is it?) Perhaps add some comments to explain the logic here.